### PR TITLE
mpi: Towards compatibility for mode!=basic and GPU

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

**Review by georgebisbas — APPROVED** [review]

File: devito/types/dense.py [hunk 1]
```diff
@@ -1610,3 +1610,30 @@ def _arg_values(self, **kwargs):
...
+
+    is_AliasFunction = True
+
+    __indices_setup__ = Function.__indices_setup__
+    __shape_setup__ = Function.__shape_setup__
+
+    @property
+    def data(self):
+        # Any attempt at allocating data by the user should fail miserably
+        raise TypeError("TempFunction cannot allocate data")
```
Comment (mloubout) [inline devito/types/dense.py:1634]: AliasFunction

**Review by mloubout — COMMENTED** [review]
What's with the imports changes?

**Comment by FabioLuporini** [comment]
> What's with the imports changes?

circular dependency otherwise mpi->types->mpi IIRC

**Review by mloubout — APPROVED** [review]
Looks good.

Should we make `Alias/Temp` internal classes `_Alias/_Temp` just a though may be a pain to update.

**Comment by FabioLuporini** [comment]
> Should we make Alias/Temp internal classes _Alias/_Temp just a though may be a pain to update.

I think you're right. I'm merging, but I'm gonna fix this now in #1703 . Not with the underscore, but by removing them from the `__all__`

**Comment by FabioLuporini** [comment]
ah, no, I cannot merge because nexu is offline. So will fix it here now

**Comment by FabioLuporini** [comment]
actually, that applies to AliasFunction, but not so sure about TempFunction. So I'll pull off AliasFunction for now

**Pull request merged**